Semantic highlighting improvements

0dinD opened the pull request:
Implements redhat-developer/vscode-java#1488
Fixes redhat-developer/vscode-java#1514

## Changes
- Added some javadoc comments to the semantic highlighting classes
- Fixed indentation of the semantic highlighting classes, as per the `.editorconfig`

#### Token types and modifiers
- Both token types and modifiers can now be found in their respective enum.
- Made the naming scheme more consistent
  - The enum members now have "Java-names", such as `TokenType.METHOD` instead of `TokenType.FUNCTION`
  - The `toString()` method on the enum members return the more generic name used in the legend.

- Removed unused token types
- Added new token types:
  - Class
  - Interface
  - Enum
  - Enum member
  - Type parameter
  - Parameter
  - Annotation member (Example: `value` in `@SuppressWarnings(value = "all")`)
- Added new token modifiers:
  - Declaration

#### Semantic tokens visitor
- Moved all the logic that figures out applicable token types and modifiers into `TokenType` and `TokenModifier` enums.
- No longer merges the entire qualified name of a package into one token
  - This is to avoid inconsistencies with the "dots", for example:
    - The dot between `java` and `lang` in `java.lang.Math` is part of a `namespace` token
    - The dot between `lang` and `Math` is not part of a `namespace` token
    - It does not make sense for any of the dots to be part of a `namespace` token
  - I am aware that this change was implemented to fix redhat-developer/vscode-java#1434
    - But in my testing this issue seems to no longer appear even with my revert
    - If someone finds that this change introduces the issue again, we can revert it
- Now tries to more precisely highlight the invocation of a constructor, fixes redhat-developer/vscode-java#1514
- Now applies token modifiers in cases where it previously did not. (Types and annotations, for example.)

## Testing
I did not manage to run the tests, since they failed halfway through even before I made any changes to the code (I will try to debug this soon). At least one test will fail because of the change I made to the merging of package names, so that test will have to be rewritten. Also, I haven't written any tests for the above changes, since I could not set up testing to work on my machine. It would be very kind of someone to help me write these tests, maybe @Eskibear since you wrote the other semantic highlighting tests? I will try to get testing to work for me in the meantime.

genie-ls commented:
Can one of the admins verify this patch?

fbricon commented:
add to whitelist

0dinD commented:
Most of the tests for semantic tokens are failing now because they are expecting the token type `type`, but are getting the more specific `class`, `interface` or `enum` token types. This is fixable by just testing for the correct token type.

A related issue to this is that color themes using the old `type` token type will no longer work correctly. This can be fixed by giving the more specific token types a supertype of `type`. I don't know if it can be done via the semantic tokens API, but one solution is to declare the semantic token types in the contribution point "semanticTokenTypes" in the package.json of vscode-java.

See https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification

fbricon commented:
@0dinD thanks for your contribution! Please make sure your commits are signed (git commit -s) and sign the Eclipse Contributor Agreement, before we can accept this PR.

For the tests, we're aware of some unstable tests that might case some build failures now and again. Now, given the scope of your changes, you can safely simply amend and run the [SemanticTokensCommandTest](https://ci.eclipse.org/ls/job/jdt-ls-pr/2169/testReport/) (`mvn clean verify -Dtest=SemanticTokensCommandTest`).

I'll let @Eskibear do the review for this one.

0dinD commented:
Thank you for the suggestion about the tests! I have now fixed the existing tests to make them run, but I think some new tests should be added because of the additions I made. 

I will look into signing the Eclipse Contributor Agreement and signing my commits.

Eskibear commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensVisitor.java:
```diff
@@ -18,204 +18,212 @@
+
+	@Override
+	public boolean visit(SimpleName node) {
+		TokenType tokenType = TokenType.getApplicableToken(node.resolveBinding());
```
https://github.com/redhat-developer/vscode-java/issues/1434 is introduced again. And by the way it's not only for `java`. 
![image](https://user-images.githubusercontent.com/2351748/87116406-5388f400-c2a8-11ea-9e8d-d0b3d17d767b.png)


![image](https://user-images.githubusercontent.com/2351748/87116297-0573f080-c2a8-11ea-9251-b82db81c2b52.png)

0dinD commented:
As far as redhat-developer/vscode-java#1434 goes, I think it would be better to fix whatever causes the binding of the package to be null, rather than working around the fact that it is null. Can you provide me with the environment and some instructions to reproduce this issue? I am currently unable to debug it since I cannot reproduce it. 

I will get to work adding new test cases for my changes.

Eskibear commented:
> Can you provide me with the environment and some instructions to reproduce this issue? I am currently unable to debug it since I cannot reproduce it.

I'm using a mac, and I have JDK 8 and 14, the issue occurs for both JDKs. The project is a simple folder without any project management tool. You can create an empty folder, open it in vscode, and create a .java file under the folder. 

```
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment Zulu14.28+21-CA (build 14.0.1+8)
OpenJDK 64-Bit Server VM Zulu14.28+21-CA (build 14.0.1+8, mixed mode, sharing)

java version "1.8.0_191"
Java(TM) SE Runtime Environment (build 1.8.0_191-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)

```

I also observed one thing (not sure if it's related), if you get an error for `import <package-name>.*`, then the package name has no binding (not correctly highlighted). 
![image](https://user-images.githubusercontent.com/2351748/87156089-b1402f00-c2ee-11ea-8137-ffaeb88135ca.png)


When I switch to Java 8, there is a warning instead of the error, but the highligting is still broken.

0dinD commented:
After some testing, I think I can conclude that this issue has something to do with the version of Java used to launch the server.
I am using Linux (Debian 10), and tried to reproduce this with your instructions using AdoptOpenJDK 8, 11 and 14 (hotspot).

#### Using JDK8 it works like expected:
![semantic-tokens-1](https://user-images.githubusercontent.com/52179873/87171675-4c2b1000-c2d3-11ea-869e-fb56073de6e8.png)
![semantic-tokens-2](https://user-images.githubusercontent.com/52179873/87171704-5947ff00-c2d3-11ea-9af1-30a5766bbe0d.png)
![semantic-tokens-3](https://user-images.githubusercontent.com/52179873/87171720-5fd67680-c2d3-11ea-8535-d96e6f6ca963.png)

#### Using JDK11 or JDK14 I get the same issue as you do:
![semantic-tokens-4](https://user-images.githubusercontent.com/52179873/87171799-7ed50880-c2d3-11ea-8ffc-f41112386c79.png)

> When I switch to Java 8, there is a warning instead of the error, but the highligting is still broken.

Are you sure that Java 8 is used to start the server? As in, `"java.home": "/path/to/jdk-8/"` in `settings.json`.

Can you confirm my results on your machine? Any ideas as to why this might be happening?
@fbricon @Eskibear

0dinD commented:
After thinking about it, I have a strong suspicion that this issue has something to do with the module system introduced in Java 9. That would explain why it does not appear on JDK8, and why it seems to only happen with the Java standard library. I will try to find out if this is the case, and what can be done about it.

0dinD commented:
I have now found an easier way to reproduce this issue. It is not necessarily dependent upon the JDK version used to launch the server, as I previously thought. Rather, the issue is introduced whenever the property `org.eclipse.jdt.core.compiler.compliance` is set to `9` or higher in the file `<workspace>/.settings/org.eclipse.jdt.core.prefs`.

This pretty much confirms my suspicion of Java 9 modules being the cause of this issue. Now it's just a matter of figuring out why `SimpleName#resolveBinding()` returns null for some simple names that are linked to Java 9 modules. Does anyone have an idea?

Also, can you verify that this issue is reproduced by setting `org.eclipse.jdt.core.compiler.compliance` to `9` or higher? (And also that it isn't an issue if the setting is `1.8` or lower.)

0dinD commented:
I have now finally found the root cause of this issue!

### The actual issue
`SimpleName#resolveBinding()` returns null for many packages in the Java standard library. The expected result is a package binding, which is needed in order to add the correct semantic token type. The lack of a package binding can be worked around, like in the previous fix, but this causes other problems, as noted in my first comment.

### Why can the package binding not be resolved?
Because it is a `ProblemPackageBinding`, which means that it will never be returned by `DefaultBindingResolver#getPackageBinding()`.

### Why is it a `ProblemPackageBinding`?
Because the package is not accessible. It is not accessible because it is not explicitly exported by the module it is contained in. In the example below, the `java` package is not exported by any module in the Java standard library, so it cannot be accessed. This does not mean that sub-packages like `java.awt` cannot be accessed, it only applies to the `java` package.
![not-accessible](https://user-images.githubusercontent.com/52179873/87201750-7f839400-c2ff-11ea-8e2e-66c0938f0667.png)

This explains why some simple names in import statements cannot have their package binding resolved: those packages are not accessible. The example below might clarify which simple names can get their package bindings resolved:
![not-accessible-2](https://user-images.githubusercontent.com/52179873/87202069-1b150480-c300-11ea-86b2-dafcd231d5f6.png)

### The current vs expected behavior
Looking at the above screenshot, the import statement of lines 3, 6 and 7 do not have any package bindings. This is probably correct since they are completely invalid import statements anyway. (In Java 9+)

However, the import statements of lines 4 and 8 contain packages with missing package bindings (hence the missing semantic tokens). One these lines, one would expect package bindings to be available since the import statements are valid. But they are not, since the packages `java`, `org` and `org.ietf` are not explicitly exported from any modules in the Java standard library.

#### To clarify the expected behavior:
A package binding should not be classified as a `ProblemPackageBinding` ([see this line](https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/tree/org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/CompilationUnitScope.java#n569)) unless the package itself is the target of an invalid import statement. 

### What can be done about this?
We can either continue to work around this issue, or get it fixed. The problem is of course that the fix needs to be made in `eclipse.jdt.core`, not in `eclipse.jdt.ls`.

What do you think is the best course of action @fbricon ? Maybe you can talk to someone at Eclipse about the issue to get it fixed? Or is it not worth it?

0dinD commented:
After making the root cause of the issue more clear, I realised that there is a better way to work around the null package bindings (See commit above). I think the logic behind it should be clear by looking at the small code addition and my javadoc comment. 

#### This is how the semantic highlighting looks now:
![not-accessible-3](https://user-images.githubusercontent.com/52179873/87214218-1580e400-c32b-11ea-9b98-f502a4256896.png)

Can you review this change? @Eskibear

fbricon commented:
@0dinD in order for us to be able to merge this PR, make sure you squash all your commits into one, and sign it off using `git commit -s`, you committer email should be identical to the one used for signing the ECA.

fbricon commented:
Even if you have unresolved packages, they should still be highlighted as packages. Same should go for java.* and others, for consistency.

<img width="149" alt="Screen Shot 2020-07-15 at 11 32 05 AM" src="https://user-images.githubusercontent.com/148698/87529247-01244a80-c68f-11ea-8e77-4f70f81ec02b.png">

0dinD commented:
@fbricon I will make sure to squash all my commits and sign once the code seems ready to merge. 

As for your second comment, I do agree, but this behavior is not something that I changed with 7b53dce . Even with the current version of eclipse.jdt.ls, semantic tokens are missing for imports where the package bindings are null as a result of package accessibility for Java 9 modules. The merging of the qualified name for the package name still only happens if the imported package is not a `ProblemPackageBinding`. 

#### This screenshot is taken with the marketplace version of vscode-java
![packages](https://user-images.githubusercontent.com/52179873/87540355-d2fb3680-c69f-11ea-9982-7a735401a99a.png)
This only happens with `org.eclipse.jdt.core.compiler.compliance` set to `9` or higher. As I said earlier, this needs to be fixed in `eclipse.jdt.core`, or else we need to create yet another workaround. 

The image you sent is another strange issue, but is actually unrelated to the package bindings, it seems. I can reproduce the same issue on the marketplace version of vscode-java. The semantic tokens on `dont` and `exists` are `type` (or `class` in this PR), which is strange. I can look into fixing this issue in this PR if you would like.

Eskibear commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensVisitor.java:
```diff
@@ -127,6 +129,52 @@ private void addToken(ASTNode node, TokenType tokenType, TokenModifier[] modifie
+
+	@Override
+	public boolean visit(QualifiedName node) {
+		if (node.resolveBinding() instanceof IPackageBinding) {
```
Can add `|| node.resolveBinding() instanceof ITypeBinding` to cover more cases. E.g.

```java
import org.jcp.*;
import org.ietf.Eskibear;
```

Eskibear commented:
I just verified your latest commit 7b53dce :
- it **does** fix the regression #1434 , e.g. `import java.util.*`
- it **does not** cover the new issue we found here, e.g.  a) `import org.ietf.Eskibear`, b) `import java.*`
 
#### Room to improve

See below snippet:
![image](https://user-images.githubusercontent.com/2351748/87622352-870ac900-c755-11ea-9172-8c4a84c1de33.png)

For **a)** 
we can at least cover cases in Line 6 & 7, by applying the same workaround to `ITypeBinding` . 
![image](https://user-images.githubusercontent.com/2351748/87622220-37c49880-c755-11ea-95fa-03b7899e0983.png)
![image](https://user-images.githubusercontent.com/2351748/87622249-490da500-c755-11ea-9aef-c5bf6c2a0c5a.png)

For **b)** 
now what I can think of is, to check whether its parent is a `ImportDeclaration`.  E.g. for QualifiedName `org.xml`,
```
> node
QualifiedName@584 "org.xml"
> node.getParent()
ImportDeclaration@586 "import org.xml.*;"
```

But it doesn't fix `import java.*` because "java" is a `SimpleName`. 

#### Alternative implementation
So here comes an alternative approach. As the issues are all about import statements, can we just do it in override `visit(ImportDeclaration)`?

0dinD commented:
Yes, overriding `visit(ImportDeclaration)` was an idea that I had. If we want to give semantic tokens to `import java.*;` then I don't think there is any other way to do it, since there are no bindings that can be resolved in that import statement. I will take a look at implementing a new workaround based on that solution.

0dinD commented:
After thinking about this some more, I realised that we do still need to visit qualified names, since package names not only occur in import statements, but also in cases where Types have their package explicitly specified:
```java
// Stupid example, but you get the point
java.lang.String string = new java.lang.String();
```

The main problem here is that we are doing guesswork as far as applying the package tokens, since they are discarded by `DefaultBindingResolver#getPackageBinding()` ([see this line](https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/tree/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/DefaultBindingResolver.java#n327)). I have no idea whether that is the intended behavior, but it does make life harder for us. 

Despite this, I made yet another workaround that should fix the comments you have made about the package highlighting. Can you review my latest commit @Eskibear ?

Eskibear commented:
Great. I tried you PR, it manages to cover all the cases we mentioned above. 
- On-demand (star) imports. e.g `import java.*`, `import org.xml.*`.   Coverred by overriding `visit(ImportDeclaration)` + `isOnDemand()` check.   
- Type imports. e.g. `import java.util.List`. Coverred by overriding `visit(QualifiedName)`

So overall it looks good to me.



> I realised that we do still need to visit qualified names, since package names not only occur in import statements, but also in cases where Types have their package explicitly specified

Not true, the problems only occur in import statements. `resolveBinding` works well for fully qualified name (e.g. `java.lang.String aString = new java.lang.String()` ) in the body. 


So my original proposal was to only override `visit(ImportDeclaration)`, roughly as below:
```java
@Override
public boolean visit(ImportDeclaration node) {
  if node.isOnDemand()
      visitPackage(node.getName())   // to highlight "java" in "import java.util.*"
  else
      visitPackageWithType(node.getName()) // to highlight "java" in "import java.util.List"
}

// in "visitPackageWithType", you can assign correct tokens for Qualifier and Name . 
```
An advantage of this proposal would be better performance. It only affects import statements, thus you don't have to perform the check `isPackageName` (where `resolveBinding()` is costly) for any other qualifiled names in the method body.

fbricon commented:
I'm planning to perform a release on Wednesday, so if we can get that in, that'd be great. @0dinD make sure you get your signed commits + ECA tomorrow at the latest.

0dinD commented:
Okay @fbricon , I can probably commit some final touches (addressing Eskibear's comment + adding new test cases) later today. I will make sure to squash and sign everything so you can merge it tomorrow if it looks ready to you. 

By the way, these changes might break styling for projects not using the new, more specific token types (Class, Interface, Enum, Parameter etc). For users of vscode-java this should not be a problem if we declare token types and their supertypes in the package.json contribution point. I plan to make a pull request for that also later today, which you should probably take a look at before you use the new version of eclipse.jdt.ls in vscode-java. For other projects using eclipse.jdt.ls, they probably have to figure out token supertypes in their own implementation, since what I am talking about is only available to VS Code extensions. 

This is the token type hierarchy I was thinking of declaring:
- `namespace`
- `type`
  - `class`
  - `interface`
  - `enum`
  - `annotation`
  - `typeParameter`
- `member`
  - `function`
    - `annotationMember`
  - `property`
    - `enumMember`
- `variable`
  - `parameter`

Do you agree? I am unsure of whether or not to give `annotationMember` a supertype of `function` or `property`, since they are technically methods but probably not often thought of as such when using them in an annotation.

0dinD commented:
I have now addressed Eskibear's comment in 3473f16. The new implementation only performs a check for null package bindings in import statements, when absolutely necessary. Also, I added some new test cases (483c0ad) to cover the new token types and modifiers I added in this PR. @Eskibear can you do a review of these changes?

Finally, I have signed the ECA and squashed my old commits into a new signed commit. So feel free to merge this if you feel that it is ready, @fbricon . Also don't forget to merge my PR redhat-developer/vscode-java#1537 for declaring token supertypes in the vscode-java extension if you merge this PR, since it ensures backwards compatibility for VS Code themes using the "old" token types.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensVisitor.java:
```diff
@@ -9,213 +9,322 @@
+		return new SemanticTokens(data);
+	}
+
+	private List<Integer> encodeTokens() {
```
encodedTokens

0dinD replied:
Fixed and amended to previous commit

Eskibear commented:
LGTM

fbricon commented:
@0dinD awesome work, thanks!

tommai78101 commented:
@0dinD @fbricon Is it me or, after the merge, the formatting is no longer consistent with the rest of the Eclipse code format?

For example, in `org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokenManager.java`, the first indentation is a 1-space wide indent, but in the original, it's a 4-space indent.

0dinD commented:
@tommai78101 Are you sure that it is a space and not a tab character that you have configured to be 1-space wide? I re-indented everything under `org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/` because the `.editorconfig` said to use tabs for indentation, but those source files were indented using spaces.

tommai78101 commented:
@0dinD Ok, so it is really me. I was just looking at this directly, and didn't double-check to see the 1-width character is a tab or not:

https://github.com/eclipse/eclipse.jdt.ls/pull/1501/commits/785a6f175ea7823dee6f177517dda4160325bb67

Thanks.
